android-components/components/feature/addons/src/main/java/mozilla/components/feature/addons/amo/AMOAddonsProvider.kt

```diff
@@ -173,26 +187,53 @@ class AMOAddonsProvider(
     }
 
     /**
-     * Fetches given Addon icon from the url and returns a decoded Bitmap
-     * @throws IOException if the request could not be executed due to cancellation,
-     * a connectivity problem or a timeout.
+     * Asynchronously loads add-on icon for the given [iconUrl] and stores in the cache.
      */
-    @Throws(IOException::class)
-    override suspend fun getAddonIconBitmap(addon: Addon): Bitmap? {
-        var bitmap: Bitmap? = null
-        if (addon.iconUrl != "") {
-            client.fetch(
-                Request(url = addon.iconUrl.sanitizeURL()),
-            ).use { response ->
-                if (response.isSuccess) {
-                    response.body.useStream {
-                        bitmap = BitmapFactory.decodeStream(it)
+    @VisibleForTesting
+    internal fun loadIconAsync(addonId: String, iconUrl: String): Deferred<Bitmap?> = scope.async {
+        val cachedIcon = iconsCache[addonId]
+        if (cachedIcon != null) {
+            logger.info("Icon for $addonId was found in the cache")
+            cachedIcon
+        } else if (iconUrl.isBlank()) {
+            logger.info("Unable to find the icon for $addonId blank iconUrl")
+            null
+        } else {
+            try {
+                logger.info("Trying to fetch the icon for $addonId from the network")
+                client.fetch(Request(url = iconUrl.sanitizeURL(), useCaches = true))
+                    .use { response ->
+                        if (response.isSuccess) {
+                            response.body.useStream {
+                                val icon = BitmapFactory.decodeStream(it)
+                                logger.info("Icon for $addonId fetched from the network")
+                                iconsCache[addonId] = icon
+                                icon
+                            }
+                        } else {
+                            // There was an network error and we couldn't fetch the icon.
+                            logger.info("Unable to fetch the icon for $addonId HTTP code ${response.status}")
+                            null
+                        }
                     }
-                }
+            } catch (e: IOException) {
+                logger.error("Attempt to fetch the $addonId icon failed", e)
+                null
             }
         }
+    }
 
-        return bitmap
+    @VisibleForTesting
+    internal suspend fun List<Addon>.loadIcons(): List<Addon> {
+        this.map {
+            // Instead of loading icons one by one
+            // lets load them async, so we can do multiple request at the time
```

Comment:
nit:

```suggestion
            // Instead of loading icons one by one, let's load them async
            // so we can do multiple request at the time.
```

Reply:
Addressed!